### ISSUE TITLE
Release 1.1 Breaks Visual Studio Intellisense

### ISSUE THREAD

**Issue by jtrahan** [cowboy/jquery-bbq#1]
Using Visual Studio 2008 and jQuery 1.3.2. I'm getting an error stating 

"
Warning 1   Error updating JScript IntelliSense: \Scripts\jquery.bbq.js: 'undefined' is null or not an object @ 786:2
"

**Comment by cowboy** [comment]
Do you get a similar (or the same) error if you try to include just my jquery-hashchange plugin? I've never used intellisense, so I don't know what the problem might be.

**Comment by jtrahan** [comment]
It is complaining about this section. If I comment out the section the intellisense quits complaining. But I don't know what all it will do when I do that since I'm only interested in using just the deparam piece.

<pre>
jq_event_special[ str_hashchange ] = $.extend( jq_event_special[ str_hashchange ], {
    

    // Augmenting the event object with the .fragment property and .getState
    // method requires jQuery 1.4 or newer. Note: with 1.3.2, everything will
    // work, but the event won't be augmented)
    add: function( handler, data, namespaces ) {
      return function(e) {
        // e.fragment is set to the value of location.hash (with any leading #
        // removed) at the time the event is triggered.
        var hash = e[ str_fragment ] = jq_param_fragment();
        
        // e.getState() works just like $.bbq.getState(), but uses the
        // e.fragment property stored on the event object.
        e.getState = function( key, coerce ) {
          return key === undefined || typeof key === 'boolean'
            ? jq_deparam( hash, key ) // 'key' really means 'coerce' here
            : jq_deparam( hash, coerce )[ key ];
        };
        
        handler.apply( this, arguments );
      };
    }
    
  });
</pre>

**Comment by cowboy** [comment]
So I need to find out which line is actually causing the problem. What if you change all that to just:

```
jq_event_special[ str_hashchange ] = $.extend( jq_event_special[ str_hashchange ], {} );
```

Does it still error?

**Comment by jtrahan** [comment]
It also complains about this piece too once I comment out the first section. Maybe the way the extend is constructed it doesn't like.

<pre>
    jq_event_special[str_hashchange] = $.extend(jq_event_special[str_hashchange], {

        // Called only when the first 'hashchange' event is bound to window.
        setup: function() {
            // If window.onhashchange is supported natively, there's nothing to do..
            if (supports_onhashchange) { return false; }

            // Otherwise, we need to create our own. And we don't want to call this
            // until the user binds to the event, just in case they never do, since it
            // will create a polling loop and possibly even a hidden IFRAME.
            fake_onhashchange.start();
        },

        // Called only when the last 'hashchange' event is unbound from window.
        teardown: function() {
            // If window.onhashchange is supported natively, there's nothing to do..
            if (supports_onhashchange) { return false; }

            // Otherwise, we need to stop ours (if possible).
            fake_onhashchange.stop();
        }

    });
</pre>

**Comment by cowboy** [comment]
So, if it is an $.extend error, we need to figure out what it is. If you replace the problematic code with either of these, is there still an error?

```
jq_event_special[ str_hashchange ] = $.extend( jq_event_special[ str_hashchange ], {a:1} );
```

or

```
jq_event_special[ str_hashchange ] = $.extend( undefined, {a:1} );
```

or

```
jq_event_special[ str_hashchange ] = $.extend( {}, {a:1} );
```

**Comment by jtrahan** [comment]
1st one gives me the undefined error. the other 2 says object required. I'm also playing around with the syntax to figure out why it is giving an error.

**Comment by jtrahan** [comment]
OK I was able to make the errors go away, however I don't know if it broke anything. 
But I changed both instances of

<pre>
jq_event_special = $.event.special
</pre>


to

<pre>
jq_event_special = $.event.special || {},
</pre>

**Comment by cowboy** [comment]
That "fix" doesn't make sense to me, because in jQuery 1.3.2, $.event.special is defined. Are you sure you're using jQuery 1.3.2?

**Comment by jtrahan** [comment]
<pre>
/*
 * jQuery JavaScript Library v1.3.2
 * http://jquery.com/
 *
 * Copyright (c) 2009 John Resig
 * Dual licensed under the MIT and GPL licenses.
 * http://docs.jquery.com/License
 *
 * Date: 2009-02-19 17:34:21 -0500 (Thu, 19 Feb 2009)
 * Revision: 6246
 */
</pre>

**Comment by jtrahan** [comment]
I also removed all scripts except for jquery and bbq just to make sure I'm not getting a weird conflict and it still complained about those pieces until I had "|| {}" added. I'm not sure how visual studio is processing the jscripts, but maybe it has some issue with the syntax. Maybe it doesn't recognize event.special, even though event.special works in runtime mode. Just one of those weird misteries I guess. However if it doesn't break anything can you add it to the trunk so that other ASP.NET developers who may use this sript doesn't run into this issue.

**Comment by cowboy** [comment]
Since `$.event.special` is defined in jQuery 1.3.2, then setting `jq_event_special = $.event.special || {},` would be introducing unnecessary code into the plugin just to work around what may be a bug in IntelliSense or some jQuery-specific IntelliSense content.

And if that weren't enough, if `$.event.special` were undefined, then `jq_event_special = $.event.special || {},` wouldn't actually work.. because the methods that are supposed to be getting added to `$.event.special` would instead be added to the `{}` object that isn't actually referenced anywhere outside the plugin.

I'd have to do `jq_event_special = $.event.special || ( $.event.special = {}),` which, if the previous code was too much, then this would be WAY too much to work around what I'm guessing is an IntelliSense-specific issue.

**Comment by jtrahan** [comment]
It looks like it may be an intellisense issue. Even with jQuery by itself, Visual Studio is not detecting event.special. only event. I guess what i'll do for now is just copy just the deparam piece of the code as this is all I actually need to parse out the data returned from your postMessage plugin ;)

**Comment by jtrahan** [comment]
AH HA!!!. Figured it out. Apparently when you use a "-vsdoc.js" file it ignores the original jscript file. So in this case the jquery-vsdoc.js I had was missing event.special. So intellisense was thinking that the object didn't exist. I noticed that the copy I had was not the most recent so right after I updated that one file the errors went away. 

So in other words you don't need to make any changes. ;) Sorry about all the confusion.
